File: number_theory/02_sieve.cpp
#include<iostream>
#include<vector>
using namespace std;

//print primes in range [1,n]
void sieve(int n){
	if( n <= 1 ){
		cout << "No primes in this range." << endl;
		return;
	}

	vector<bool> primes(n+1,true);
	primes[0] = false;
	primes[1] = false;

	for( int i = 2; i*i <= n; i++ ){
		if( !primes[i] ) continue;
		for( int j = i*i; j <= n; j += i )
			primes[j] = false;
	}

	for( int i = 0; i <= n; i++ )
		if( primes[i] )
			cout << i << " ";

	cout << endl;
}

int main(){
	int n;
	cout << "find primes upto : ";
	cin >> n;
	sieve(n);
	return 0;
}

File: number_theory/02_sieve.py

def sieve(n : int) -> list:
	if( n == 0 ):
		return [False]
	if( n == 1 ):
		return [False,False]

	primes = [True for i in range(0,n+1)]
	primes[0] = False
	primes[1] = False

	i = 2
	while( i*i <= n ):
		if( not primes[i] ):
			i += 1
			continue
		for j in range(i*i, n+1, i):
			primes[j] = False
		i += 1

	return primes



n = int(input("find primes upto : "))
primes = sieve(n)
for i,val in enumerate(primes):
	if(val):
		print(i, end=' ')
print()